[Term Entry - Edit]: Python `namedtuple` (#7551)
Codecademy/docs: content/python/concepts/collections-module/terms/namedtuple/namedtuple.md

  - 'learn-python-3'
  - 'paths/computer-science'
---

A **`namedtuple`** is a factory function in Python's `collections` module that creates [`tuple`](https://www.codecademy.com/resources/docs/python/tuples) subclasses with named fields. It provides an immutable data structure that allows access to elements through descriptive field names instead of numeric indices, making Python `namedtuple` structures more readable and self-documenting while maintaining the memory efficiency and performance characteristics of regular tuples.

## Syntax

```pseudo
namedtuple(typename, field_names, *, rename=False, defaults=None, module=None)
```

**Parameters:**

- `typename`: A string specifying the name of the new tuple subclass
- `field_names`: A sequence of strings or a single string with field names separated by whitespace and/or commas
- `rename`: If `True`, invalid field names are automatically replaced with positional names
- `defaults`: An iterable of default values for the rightmost parameters
- `module`: Value for the `__module__` field

**Return value:**

A new tuple subclass with named fields that can be accessed via dot notation.

## Example 1: Python `namedtuple` Basic Example

> This example demonstrates how to create a Python `namedtuple`, access its elements, and perform basic operations with named tuples Python:

Reviewer: ```suggestion
This example demonstrates how to create a Python `namedtuple`, access its elements, and perform basic operations in Python:
```

coord2 = Coordinate._make(point_tuple)

print(coord1)
print(coord2)
```

This example results in the following output:

```shell
Coordinate(x=10, y=20, z=30)
Coordinate(x=5, y=15, z=25)
```

### Example 3: Python Namedtuple `_asdict()` Method

Converts a `namedtuple` instance to an OrderedDict, preserving the field names as keys and their corresponding values:

```py
from collections import namedtuple

# Create namedtuple
Student = namedtuple('Student', ['name', 'grade', 'subject'])
student = Student('Alice Johnson', 95, 'Mathematics')

# Convert to dictionary
student_dict = student._asdict()
print("As dictionary:", student_dict)

# Access dictionary values
print("Name from dict:", student_dict['name'])
print("Grade from dict:", student_dict['grade'])
```

This example results in the following output:

```shell
As dictionary: OrderedDict([('name', 'Alice Johnson'), ('grade', 95), ('subject', 'Mathematics')])
Name from dict: Alice Johnson
Grade from dict: 95
```

### Example 4: Namedtuple Using `**` (Double Star) Operator

Unpacks a dictionary or `namedtuple` to create new `namedtuple` instances using keyword arguments:

```py
from collections import namedtuple

> # Create namedtuple class
> Product = namedtuple('Product', ['name', 'price', 'category'])
> 
> # Create from dictionary using ** operator
> product_data = {
>   'name': 'Laptop',
>   'price': 999.99,
>   'category': 'Electronics'
> }
> 
> laptop = Product(**product_data)
> print("Created from dict:", laptop)
> 
> # Create from existing namedtuple using ** operator
> existing_product = Product('Phone', 599.99, 'Electronics')
> modified_product = Product(**existing_product._asdict(), price=549.99)
> print("Modified product:", modified_product)

Reviewer: ```suggestion
# Create namedtuple class
Product = namedtuple('Product', ['name', 'price', 'category'])

# Create from dictionary using ** operator
product_data = {
    'name': 'Laptop',
    'price': 999.99,
    'category': 'Electronics'
}

laptop = Product(**product_data)
print("Created from dict:", laptop)

# Create from existing namedtuple using _replace (cleanest way)
existing_product = Product('Phone', 599.99, 'Electronics')
modified_product = existing_product._replace(price=549.99)
print("Modified product:", modified_product)
```
